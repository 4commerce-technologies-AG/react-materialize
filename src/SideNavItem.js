import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import UserView, { UserShape } from './UserView';

const SideNavItem = ({
  divider,
  subheader,
  userView,
  icon,
  href = '#!',
  waves,
  user = {},
  children,
  ...props
}) => {
  const itemClasses = { divider };

  const linkClasses = {
    subheader,
    'waves-effect': waves
  };

  return (
    <li className={cx(itemClasses)} {...props}>
      {userView && user && <UserView {...user} />}
      {!userView && (
        <a className={cx(linkClasses)} href={href}>
          {icon && React.cloneElement(icon, { className: 'prefix' })}
          {children}
        </a>
      )}
    </li>
  );
};

SideNavItem.propTypes = {
  children: PropTypes.node,
  divider: PropTypes.bool,
  subheader: PropTypes.bool,
  userView: PropTypes.bool,
  waves: PropTypes.bool,
  href: PropTypes.string,
  icon: PropTypes.node,
  user: PropTypes.shape(UserShape)
};

export default SideNavItem;
