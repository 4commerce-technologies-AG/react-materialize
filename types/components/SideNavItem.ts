import * as React from 'react';

export interface SideNavItemProps {
  href?: string;
  waves?: boolean;
  children?: string;
  divider?: boolean;
  subheader?: boolean;
  userView?: boolean;
  icon?: React.ReactNode;
  user?: any; // TODO use UserShape
};

/**
 * React Materialize: SideNavItem
 */
declare const SideNavItem: React.FC<SideNavItemProps>

export default SideNavItem;
