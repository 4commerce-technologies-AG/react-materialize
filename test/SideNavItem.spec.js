import React from 'react';
import { shallow } from 'enzyme';
import { assert } from 'chai';
import SideNavItem from '../src/SideNavItem';
import UserView from '../src/UserView';

function setup(props = {}, children) {
  const propsIn = {
    ...props
  };
  const component = <SideNavItem {...propsIn}>{children}</SideNavItem>;
  const wrapper = shallow(component);
  const firstChild = wrapper.childAt(0);
  const propsOut = wrapper.props();
  return {
    propsIn,
    wrapper,
    firstChild,
    propsOut
  };
}

describe('<SideNavItem />', () => {
  test('renders', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  test('should render', () => {
    const { wrapper, firstChild } = setup();
    assert(wrapper.length === 1, 'should render the wrapper');
    assert(firstChild.type() === 'a', 'should render the link');
    assert.equal(
      firstChild.props().href,
      '#!',
      'should render default link href'
    );
    const icon = firstChild.find('.material-icons');
    assert.equal(icon.length, 0, 'should not render an icon');
  });

  test('should render children', () => {
    const { wrapper } = setup({}, <span className="test-child" />);
    const children = wrapper.find('.test-child');
    assert(children.length === 1, 'should render given children');
  });

  test('should not render children for a userview', () => {
    const { wrapper } = setup(
      { userView: true },
      <span className="test-child" />
    );
    const children = wrapper.find('.test-child');
    assert(children.length === 0, 'should not render given children');
  });

  test('should consume props it uses', () => {
    const { propsOut } = setup({
      id: 'test123',
      divider: true,
      waves: true,
      'data-someProp': 'true'
    });
    assert.isUndefined(
      propsOut.divider,
      'should not transfer the divider prop'
    );
    assert.isUndefined(propsOut.waves, 'should not transfer the waves prop');
    assert.equal(propsOut.id, 'test123', 'should transfer the id prop');
    assert.equal(
      propsOut['data-someProp'],
      'true',
      'should transfer the data-someProp prop'
    );
  });

  test('should render a divider class when needed', () => {
    const { wrapper } = setup({
      divider: true
    });
    assert.isTrue(wrapper.hasClass('divider'), 'should render a divider class');
  });

  test('should render a subheader or a waves-effect class when needed', () => {
    const { firstChild } = setup({
      waves: true,
      subheader: true
    });
    assert.isTrue(
      firstChild.hasClass('subheader'),
      'should render a subheader class'
    );
    assert.isTrue(
      firstChild.hasClass('waves-effect'),
      'should render a waves-effect class'
    );
  });

  test('should render an icon when needed', () => {
    const { firstChild } = setup({
      icon: 'car'
    });
    const icon = firstChild.find('.material-icons');
    assert.isTrue(icon.equals(<i className="material-icons">car</i>));
  });

  test('should render a given href', () => {
    const { firstChild } = setup({
      href: 'www.test.nl'
    });
    assert.equal(
      firstChild.props().href,
      'www.test.nl',
      'should render correct link href'
    );
  });

  test('should render a userview', () => {
    const user = { name: 'test' };
    const { firstChild } = setup({
      userView: true,
      user
    });
    assert(firstChild.equals(<UserView {...user} />));
  });
});
