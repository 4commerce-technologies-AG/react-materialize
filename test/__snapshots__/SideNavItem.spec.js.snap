// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideNavItem /> renders 1`] = `
<li
  className=""
>
  <a
    className=""
    href="#!"
  />
</li>
`;
